### PR TITLE
fix: split_long_word incorrectly splits multi-codepoint character

### LINKED ISSUE #168: [Bug] split_long_word incorrectly splits multi-codepoint character
### Describe the bug

`split_long_word()` may split multi-codepoint characters, such as â (`a` + `◌̂` in [NFD](https://unicode.org/reports/tr15/#Norm_Forms)) and 🇺🇳, in the middle.

### Steps to reproduce

```rust
use comfy_table::utils::formatting::content_split::split_long_word;
use unicode_width::UnicodeWidthStr;

fn main() {
    let emoji = "🙂‍↕️"; // U+1F642 U+200D U+2195 U+FE0F head shaking vertically
    assert_eq!(emoji.len(), 13);
    assert_eq!(emoji.chars().count(), 4);
    assert_eq!(emoji.width(), 2);

    let (word, remaining) = split_long_word(emoji.width(), &emoji);

    println!("{word}");
    assert_eq!(word, "\u{1F642}\u{200D}");
    assert_eq!(word.len(), 7);
    assert_eq!(word.chars().count(), 2);
    assert_eq!(word.width(), 2);

    println!("{remaining}");
    assert_eq!(remaining, "\u{2195}\u{FE0F}");
    assert_eq!(remaining.len(), 6);
    assert_eq!(remaining.chars().count(), 2);
    assert_eq!(remaining.width(), 2);
}
```

> Image

### Logs (if applicable)

_No response_

### Operating system

macOS 15.2

### Comfy-table version

v7.1.3

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tisonkun**
This closes #168.

cc @Nukesor @uasi

### CODE REVIEW

File: src/utils/formatting/content_split/normal.rs [hunk 1]
```diff
@@ -22,12 +23,12 @@ pub fn split_long_word(allowed_width: usize, word: &str) -> (String, String) {
...
     let mut parts = String::new();
 
-    let mut char_iter = word.chars().peekable();
+    let mut char_iter = word.graphemes(true).peekable();
     // Check if the string might be too long, one character at a time.
     // Peek into the next char and check the exit condition.
     // That is, pushing the next character would result in the string being too long.
     while let Some(c) = char_iter.peek() {
-        if (current_width + c.width().unwrap_or(1)) > allowed_width {
+        if (current_width + c.width()) > allowed_width {
```
Comment (tisonkun) [inline src/utils/formatting/content_split/normal.rs:31]: I'm considering if we'd use `width_cjk` instead of `width` here.

cc @Manishearth @Jules-Bertholet any downside of `width_cjk` or would it introduces some typically unexpected manner compared to `width`?

**Comment by tisonkun** [comment]
This patch doesn't fix `split_long_word` under the feature flag `custom_styling`.

At there we use `console::AnsiCodeIterator` that takes a `&str` directly and cannot be composed with `UnicodeSegmentation`. Perhaps need extra investigation.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -36,6 +36,11 @@ jobs:
           - target: wasm32-wasi
             os: ubuntu-latest
             minimal_setup: true
+            toolchain: "1.74"
+          - target: wasm32-wasip1
+            os: ubuntu-latest
+            minimal_setup: true
+            toolchain: "stable"
```
Comment (tisonkun) [inline .github/workflows/test.yml:43]: According to https://blog.rust-lang.org/2024/04/09/updates-to-rusts-wasi-targets.html.

File: src/utils/formatting/content_split/normal.rs (hunk 1 above)
Comment (Jules-Bertholet) [reply src/utils/formatting/content_split/normal.rs:31]: CJK will sometimes give longer (never shorter) widths than non-CJK. In the current published version of `unicode-width`, the list of affected characters is approximately (not exactly) <https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%253AEast_Asian_Width%253DA%253A%5D-%5B%253AGeneral_Category%253DLetter%253A%5D-%5B%253AGeneral_Category%253DModifier_Symbol%253A%5D>.

My recommendation would be to default to non-CJK, and then maybe add CJK width calculation as an option.

One minor note is that `unicode-width` does not guarantee that the width of a string equals the sum of the widths of its grapheme clusters. The most prominent exception to this is Arabic `لا`‎ (2 codepoints and 2 graphemes, but width of 1). However, there are currently no libraries to do the segmentation “properly”. Also, all terminals I know of render it wrong anyway. So not something you should be worried about for now, since it will look broken no matter what you do. TLDR use `.graphemes()` for now, but be aware that there are still edge cases, and in the future the best practice might change.

Comment (tisonkun) [reply src/utils/formatting/content_split/normal.rs:31]: @Jules-Bertholet Thanks for your information!

So, I'd keep using `width` until we notice certain use cases that `with_cjk` would help.

And I'd show my respect for all your efforts in unicode-rs. I have written Java, and it's quite a headache to handle many details in Unicode especially the width calculation (for alignment).

Comment (Jules-Bertholet) [reply src/utils/formatting/content_split/normal.rs:31]: Actually, looking more at your library’s docs, it seems you do try to support some amount of non-TTY usage (with `tty` feature disabled). So handling Arabic properly might be worth it for you.

In that case, one possibility is to, instead of calculating the width if each grapheme and accumulating into `current_width`, calculate the width of the entire word up to the nth grapheme each time. That is O(n<sup>2</sup>) though…

File: src/utils/formatting/content_split/normal.rs [hunk 3]
```diff
@@ -22,29 +23,24 @@ pub fn split_long_word(allowed_width: usize, word: &str) -> (String, String) {
...
 
-    let mut char_iter = word.chars().peekable();
-    // Check if the string might be too long, one character at a time.
-    // Peek into the next char and check the exit condition.
-    // That is, pushing the next character would result in the string being too long.
-    while let Some(c) = char_iter.peek() {
-        if (current_width + c.width().unwrap_or(1)) > allowed_width {
+    let mut graphmes = word.graphemes(true).peekable();
+    // Check if the string might be too long, one Unicode grapheme at a time.
+    // Peek into the next grapheme and check the exit condition.
```
Comment (Nukesor) [inline src/utils/formatting/content_split/normal.rs:28]:
```suggestion
    // Check if the string might be too long, one Unicode grapheme at a time.
    // Peek into the next grapheme and check the exit condition.
    //
    // This code uses graphemes to handle both zero-width joiner[0] UTF-8 chars, which
    // combine multiple UTF-8 chars into a single grapheme, and variant selectors [1],
    // which pick a certain variant of the preceding char.
    // 
    // [0]: https://en.wikipedia.org/wiki/Zero-width_joiner
    // [1]: https://en.wikipedia.org/wiki/Variation_Selectors_(Unicode_block)
```

File: src/utils/formatting/content_split/normal.rs (hunk 1 above)
Comment (Nukesor) [reply src/utils/formatting/content_split/normal.rs:31]: Thanks for the insights @Jules-Bertholet.
Also TIL about CJK.

Comment (Nukesor) [reply src/utils/formatting/content_split/normal.rs:31]: Regarding the `non-TTY` usage, this lib is still very much for terminal-like environments, even if they don't have tty functions to, for example, detect the current terminal width.

**Comment by Nukesor** [comment]
Heyo :) 

Thanks a lot for the research and work into this topic! Much appreciated.
My comments are mostly about testing and docs, the rest looks good to me :) 

1. Could you go ahead and add this regression test case to the `utf_8_characters.rs` integration test file:
```
/// Handle emojis that're joined via the "zero-width joiner" character U+200D and contain variant
/// selectors.
///
/// Those composite emojis should be handled as a single grapheme and thereby have their width
/// calculated based on the grapheme length instead of the individual chars.
///
/// This is also a regression test, as previously emojis were split in the middle of the joiner
/// sequence, resulting in two different emojis on different lines.
#[test]
fn zwj_utf8_word_splitting() {
    let mut table = Table::new();
    table
        .set_width(8)
        .set_content_arrangement(ContentArrangement::Dynamic)
        .set_header(vec!["test"])
        .add_row(vec!["ab🙂‍↕️def"]);

    println!("{table}");
    let expected = "
+------+
| test |
+======+
| ab🙂‍↕️ |
| def  |
+------+";
    println!("{expected}");
    assert_eq!(expected, "\n".to_string() + &table.to_string());
}
```
   Also, it would be great if you could test other edge-cases you can come up with :)
2. I tested the performance and this actually has a pretty hefty impact, the more chars are in the table (as expected).
    The benchmark for a big table went from 15ms to 25ms so a +~67% performance hit.
    
    I think this is worth it, but it's something we need to be aware of.
3. > UPDATE: It seems we can handle str_slice there. The AnsiCodeIterator is only related to handle escape characters.
    Could you write an integration test case for that as well?

**Review by Nukesor — CHANGES_REQUESTED** [review]
Heyo :) 

Thanks a lot for the research and work into this topic! Much appreciated.
My comments are mostly about testing and docs, the rest looks good to me :) 

1. Could you go ahead and add this regression test case to the `utf_8_characters.rs` integration test file:
   ```rs
   /// Handle emojis that're joined via the "zero-width joiner" character U+200D and contain variant
   /// selectors.
   ///
   /// Those composite emojis should be handled as a single grapheme and thereby have their width
   /// calculated based on the grapheme length instead of the individual chars.
   ///
   /// This is also a regression test, as previously emojis were split in the middle of the joiner
   /// sequence, resulting in two different emojis on different lines.
   #[test]
   fn zwj_utf8_word_splitting() {
       let mut table = Table::new();
       table
           .set_width(8)
           .set_content_arrangement(ContentArrangement::Dynamic)
           .set_header(vec!["test"])
           .add_row(vec!["ab🙂‍↕️def"]);

       println!("{table}");
       let expected = "
   +------+
   | test |
   +======+
   | ab🙂‍↕️ |
   | def  |
   +------+";
       println!("{expected}");
       assert_eq!(expected, "\n".to_string() + &table.to_string());
   }
   ```
   Also, it would be great if you could test other edge-cases you can come up with :)
2. I tested the performance and this actually has a pretty hefty impact, the more chars are in the table (as expected).
   The benchmark for a big table went from 15ms to 25ms so a +~67% performance hit.
   
   I think this is worth it, but it's something we need to be aware of.
3. > UPDATE: It seems we can handle str_slice there. The AnsiCodeIterator is only related to handle escape characters.

   Could you write an integration test case for that as well?

File: src/utils/formatting/content_split/normal.rs (hunk 3 above)
Comment (tisonkun) [reply src/utils/formatting/content_split/normal.rs:29]: Updated.

**Comment by tisonkun** [comment]
> Could you write an integration test case for that as well?

Let me think of it. I'm not quite familiar with escape characters, though.

File: src/utils/formatting/content_split/normal.rs (hunk 1 above)
Comment (tisonkun) [reply src/utils/formatting/content_split/normal.rs:38]: For the test case @Nukesor proposed, I noticed that `"ab🙂‍↕️def".width()` or `width_cjk` both return `7`, but it actually take more than 7 character width in terminal:

```
+---------+
| test    |
+=========+
| ab🙂‍↕️def |
+---------+
```

Not sure if unicode-rs has some functions to calculate emoji's display width "correctly" (I don't know what is a proper "correct" in this context, though).

**Comment by tisonkun** [comment]
> I tested the performance and this actually has a pretty hefty impact the more chars are in the table (as expected). The benchmark for a big table went from 15ms to 25ms so a +~67% performance hit.

This seems we pay for unicode grapheme cluster calculation :P

**Comment by Nukesor** [comment]
> Let me think of it. I'm not quite familiar with escape characters, though.

Fair enough. Just let it be if it's too much of a hassle :)

File: Cargo.toml [hunk 4]
```diff
@@ -53,15 +53,18 @@ debug = []
 integration_test = []
 
 [dependencies]
+unicode-segmentation = { version = "1" }
+unicode-width = { version = "0.2" }
+
+# Optional dependencies
```
Comment (Nukesor) [inline Cargo.toml:59]: Also please let the options ordered :)

They're auto-ordered via taplot.

**Review by Nukesor — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 4 above)
Comment (tisonkun) [reply Cargo.toml:59]: ansi-str < console in lex order so it should be ordered already?

**Comment by tisonkun** [comment]
> Fair enough. Just let it be if it's too much of a hassle :)

Yeah .. I saw your PR in #171. Perhaps we can merge this patch first and improve later.

And I just noticed that the alignment different from text and console output: there seems to be more to learn :P

**Comment by tisonkun** [comment]
@Nukesor any blocker on me?

File: src/utils/formatting/content_split/normal.rs (hunk 1 above)
Comment (Jules-Bertholet) [reply src/utils/formatting/content_split/normal.rs:38]: `🙂‍↕️` will have width 2 (looks like head shaking vertically or 4 (looks like 🙂↕️) depending on the font and terminal. It’s not consistent.

**Comment by Nukesor** [comment]
Nope :D 

Only blocker is spare time, but I just found a few hours to do chores ;) 

Went through the MR again and it looks good to me. Thanks for handling a few chores while at it! Much appreciated :)

**Review by Nukesor — APPROVED** [review]
